runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStageTest.java

```diff
@@ -0,0 +1,747 @@
+            .build();
+    PTransform parDoTransform =
+        PTransform.newBuilder()
+            .putInputs("input", "read.out")
+            .putOutputs("output", "parDo.out")
+            .setSpec(
+                FunctionSpec.newBuilder()
+                    .setUrn(PTransformTranslation.PAR_DO_TRANSFORM_URN)
+                    .setPayload(
+                        ParDoPayload.newBuilder()
+                            .setDoFn(SdkFunctionSpec.newBuilder().setEnvironmentId("common"))
+                            .build()
+                            .toByteString()))
+            .build();
+    PTransform flattenTransform =
+        PTransform.newBuilder()
+            .putInputs("readInput", "read.out")
+            .putInputs("parDoInput", "parDo.out")
+            .putOutputs("output", "flatten.out")
+            .setSpec(FunctionSpec.newBuilder().setUrn(PTransformTranslation.FLATTEN_TRANSFORM_URN))
+            .build();
+    PTransform windowTransform =
+        PTransform.newBuilder()
+            .putInputs("input", "flatten.out")
+            .putOutputs("output", "window.out")
+            .setSpec(
+                FunctionSpec.newBuilder()
+                    .setUrn(PTransformTranslation.WINDOW_TRANSFORM_URN)
+                    .setPayload(
+                        WindowIntoPayload.newBuilder()
+                            .setWindowFn(SdkFunctionSpec.newBuilder().setEnvironmentId("common"))
+                            .build()
+                            .toByteString()))
+            .build();
+
+    QueryablePipeline p =
+        QueryablePipeline.fromComponents(
+            Components.newBuilder()
+                .putTransforms("read", readTransform)
+                .putPcollections("read.out", readDotOut)
+                .putTransforms("parDo", parDoTransform)
+                .putPcollections(
+                    "parDo.out", PCollection.newBuilder().setUniqueName("parDo.out").build())
+                .putTransforms("flatten", flattenTransform)
+                .putPcollections(
+                    "flatten.out", PCollection.newBuilder().setUniqueName("flatten.out").build())
+                .putTransforms("window", windowTransform)
+                .putPcollections(
+                    "window.out", PCollection.newBuilder().setUniqueName("window.out").build())
+                .putEnvironments("common", Environment.newBuilder().setUrl("common").build())
+                .build());
+
+    ExecutableStage subgraph =
+        GreedilyFusedExecutableStage.forRootTransform(p, getOnlyElement(p.getRootTransforms()));
+    assertThat(subgraph.getOutputPCollections(), emptyIterable());
+    assertThat(subgraph.getInputPCollection().isPresent(), is(false));
+    assertThat(
+        subgraph.toPTransform().getSubtransformsList(),
+        containsInAnyOrder("read", "parDo", "flatten", "window"));
+  }
+
+  @Test
+  public void fusesFlattenWithDifferentEnvironmentInputs() {
```

Comment:
Add to this test case another ParDo from "rare" environment that consumes the output of flatten.

Reply:
Done (A & B) -> flatten -> (A' & B'), called `flattenWithHeterogeneousInputsAndOutputs`